frontend/src/pages/modelServing/screens/metrics/utils.tsx

```diff
@@ -0,0 +1,331 @@
+    ).map((xValue) => ({
+      name: 'Threshold',
+      x: xValue,
+      y: threshold,
+    })),
+    (value) => value.x,
+  );
+
+export const formatToShow = (timeframe: TimeframeTitle): 'date' | 'second' | undefined => {
+  switch (timeframe) {
+    case TimeframeTitle.ONE_HOUR:
+    case TimeframeTitle.ONE_DAY:
+      return undefined;
+    default:
+      return 'date';
+  }
+};
+
+export const per100: TranslatePoint = (point) => ({
+  ...point,
+  y: Number((point.y / 100).toFixed(2)),
+});
+
+export const toPercentage: TranslatePoint = (point) => ({
+  ...point,
+  y: point.y * 100,
+});
+
+export const createGraphMetricLine = ({
+  metric,
+  name,
+  translatePoint,
+}: NamedMetricChartLine): GraphMetricLine =>
+  metric.data?.map<GraphMetricPoint>((data) => {
+    const point: GraphMetricPoint = {
+      x: data[0] * 1000,
+      y: parseFloat(data[1]),
+      name,
+    };
+    if (translatePoint) {
+      return translatePoint(point);
+    }
+    return point;
+  }) || [];
+
+export const useStableMetrics = (
+  metricChartLine: MetricChartLine,
+  chartTitle: string,
+): NamedMetricChartLine[] => {
+  const metricsRef = React.useRef<NamedMetricChartLine[]>([]);
+
+  const metrics = Array.isArray(metricChartLine)
+    ? metricChartLine
+    : [{ ...metricChartLine, name: metricChartLine.name ?? chartTitle }];
+
+  if (
+    metrics.length !== metricsRef.current.length ||
+    metrics.some((graphLine, i) => graphLine.metric !== metricsRef.current[i].metric)
+  ) {
+    metricsRef.current = metrics;
+  }
+  return metricsRef.current;
+};
+
+export const getBreadcrumbItemComponents = (
+  breadcrumbItems: BreadcrumbItemType[],
+): ReactElement[] =>
+  breadcrumbItems.map((item) => (
+    <BreadcrumbItem
+      isActive={item.isActive}
+      key={item.label}
+      render={() => (item.link ? <Link to={item.link}>{item.label}</Link> : <>{item.label}</>)}
+    />
+  ));
+
+const checkThresholdValid = (metricType: BiasMetricType, thresholdDelta?: number): boolean => {
+  if (thresholdDelta !== undefined) {
+    if (metricType === BiasMetricType.SPD) {
+      // SPD, no limitation, valid
+      return true;
+    }
+
+    if (metricType === BiasMetricType.DIR) {
+      if (thresholdDelta >= 0 && thresholdDelta < 1) {
+        // 0<=DIR<1 , valid
+        return true;
+      }
+      // DIR, not within the range, invalid
+      return false;
+    }
+    // not SPD not DIR, undefined for now, metricType should be selected, invalid
+    return false;
+  }
+  // not input anything, invalid
+  return false;
+};
+
+const checkBatchSizeValid = (batchSize?: number): boolean => {
+  if (batchSize !== undefined) {
+    if (Number.isInteger(batchSize)) {
+      // size > 2, integer, valid
+      if (batchSize >= 2) {
+        return true;
+      }
+      // size <= 2, invalid
+      return false;
+    }
+    // not an integer, invalid
+    return false;
+  }
+  // not input anything, invalid
+  return false;
```

Comment:
Redundant return values :)

Reply:
Good spot, I've left Juntao's comments in as I think they add value - just removed the superfluous return statements